### ISSUE TITLE
The labels were leaked during the testing

### ISSUE THREAD

**Issue by Hubert161** [xiaxin1998/DHCN#2]
Thank you for your sharing.

However, we found a very serious problem in the DHCN, the labels were leaked during the testing.

As mentioned in the paper, a sequence splitting method is used to augment and label the dataset. Figure 1 shows an example of the data after preprocessing.

![image](https://user-images.githubusercontent.com/76140150/102483731-1e2ad280-40a0-11eb-9a8a-18ec52f12449.png)

Notice that, the data is not shuffled during the training and testing in DHCN, which means the last item in the previous sequence is the target item of the current sequence. (The target of each sequence is shown in Figure 2.)

![image](https://user-images.githubusercontent.com/76140150/102483848-4dd9da80-40a0-11eb-9ee0-cc450f006cf2.png)

In DHCN, you convert each batch (100 sequences) into a hypergraph, and the prediction for each sequence in the current batch is based on the items in the sequence and the hypergraph. However, previous sequences contain both the items and the target item of the current sequence, which means in the hypergraph the items of each sequence directly connect with the target item (The co-occur of items in previous sequences makes the target item have a great influence on the current prediction according to the `Equation 2` in the paper). 

In my opinion, we should not make predictions with the help of neighbor sequences in the test data, unless we shuffle the dataset or do not use the sequence splitting method to augment and label the dataset.  ([CSRM](https://github.com/wmeirui/CSRM_SIGIR2019) uses neighbor sequences to help predictions, but it shuffles the dataset before training and testing.)

This is the result of DHCN we obtained on Diginetica after data shuffling:  `Prec@20: 49.14`, `MRR@20: 16.05`

**Comment by Coder-Yu** [comment]
Thank you for identifying the problem. We will re-conduct the experiments to see what will happen. We will update the arXiv paper and the AAAI submission (before the end of the conference, the submission can be updated) if revision is needed.

**Comment by Hubert161** [comment]
Thank you for your quick response.

**Comment by Coder-Yu** [comment]
However, in my opinion, shuffling the test set is compulsory while shuffling the training set is not necessary. Because all the data in the training set is observed.

**Comment by Coder-Yu** [comment]
Anyway, we are conducting experiments with a shuffle on the test set. Let's see what will happen.

**Comment by Hubert161** [comment]
Yes, I agree with you that the shuffling of the training set is not necessary.

**Comment by Coder-Yu** [comment]
We have seen a significant drop in the performance of DHCN. Although imposing self-supervised learning on DHCN could make the performance comparable to the baselines, we think it is necessary to improve the vanilla DHCN. The revised version of codes and paper will be soon available.

**Comment by Coder-Yu** [comment]
However, it seems that it is a common issue in existing session-based recommendation methods such as SR-GNN and FGNN. The implementations of these models also do not shuffle the test set, which also leads to the same problem. We re-ran SR-GNN and also observed a performance drop. 
It seems that we should correct all these methods for a fair comparison.

**Comment by Hubert161** [comment]
Thank you for your response.

However, we think that shuffling the test set will not affect the performance of SR-GNN and FGNN. Because for SR-GNN and FGNN, the prediction for each sequence in the test set is independent and will not be affected by neighbor sequences. We believe that shuffling the test set will only change the order of predictions, but will not change the result of prediction for each sequence.

To prove our point of view, we evaluated SR-GNN on Diginetica, where we ran it 3 times in each case and averaged the results.

Here is the result of SR-GNN we obtained on Diginetica,
**without test set shuffling**: `Prec@20: 50.68`, `MRR@20: 17.48`
**with test set shuffling**: `Prec@20: 50.75`, `MRR@20: 17.43`

**Comment by Coder-Yu** [comment]
Our results are different. For SR-GNN with test set shuffling, prec@20: 49.33, MRR@20: 15.77.

We use the tensorflow version of SR-GNN released by the authors.

**Comment by Hubert161** [comment]
Our results are based on the PyTorch version of SR-GNN released by the authors. 

And we just evaluated the Tensorflow version of SR-GNN on Diginetica.  However, we observed that the performance of the Tensorflow version of SR-GNN on the Diginetica is worse than the Pytorch version.  

The authors of SR-GNN have explained the reason for the inconsistent performance of two versions in [issue#33](https://github.com/CRIPAC-DIG/SR-GNN/issues/33), the Tensorflow version is for reference only and the reported performance is based on PyTorch version.

So we strongly recommend you using the PyTorch version to obtain the results of SR-GNN.

By the way, we found that the results of SR-GNN on Diginetica are consistent no matter shuffle the test data or not, which are about `Prec@20: 49.93` and `MRR@20: 16.44`. Would you mind evaluating the Tensorflow version of SR-GNN without shuffling the test data on Diginetica now？ The comparison of results of SR-GNN with and without test data shuffling can make us quickly discover the impact of test data shuffling on SR-GNN.

Thank you for your patience and sincere response.

**Comment by Coder-Yu** [comment]
Thank you for the report. We are striving to address the identified issue of our codes. Once we manage to address it and make it better than SOTA methods, we would update the codes. Currently, the slightly modified version shows promising results. Anyway, we are sorry for the issue and the release of problematic codes.

**Comment by Coder-Yu** [comment]
Hi, the first author has addressed the issue and updated the paper and implementation. Please refer to the newest paper (the dropbox link) and the codes for the details.
